content/post/difftree.md

```diff
@@ -0,0 +1,501 @@
+---
+
+![Front image](../../static/post/difftree/intro.jpg "Similar trees next to each other.")
+
+## Introduction
+
+Do you know how Git identifies what files has changed between two commits?
+
+It is a fairly common operation you probably use every day when you review
+a pull requests, when checking your staged files... it involves prefix trees and
+Merkle trees and lots of comparisons.
+
+In this blog post I will walk you through every step in the process in an
+intuitive way; I will skip some of the hairy details so I can focus on the main
+problems and solutions, and present them in a clear and understandable way.
+
+This blog post has been inspired by my own implementation of the difftree
+algorithm as part of my work in [go-git](http://github.com/src-d/go-git), which
+in turn has been inspired by [Git](https://git-scm.com/) itself and
+[libgit2](https://libgit2.github.com/).
+
+I will use snipets of Go code here and there to illustrate some concepts, they
+are all simplified versions of the real Go code we use in go-git.
+
+## Git trees
+
+Whenever you make a Git commit, you are saving a snapshot of your project into
+your repository, this is, Git is remembering the state of all the files you have
+staged and is storing all that information in a tree structure inside the
+commit.
+
+This tree structure is composed of nodes that represent your files and
+directories.  I call this kind of tree a `MerkleTrie`, as it has properties from
+both prefix trees or tries and Merkle trees, let us get into the details!
+
+### Git trees as tries
+
+In a Git tree every node has a **name**, which is the filename portion of the
+path to the file or directory it represents, as in the figure below:
+
+!["An example of a tree with some names in their nodes."](../../static/post/difftree/names.png "A typical Git tree, with some files and directories. The names of the nodes are shown between double quotes.")
+
+The practical implications of this are:
+
+- The root has the empty string as its name.
+
+- Then name of a node is not its full path, and does not provide enough
+  information to know where in the tree the node is located.
+
+- The path of a node is the list of its ancestors, starting from the root and
+  ending in the node itself, for example the path to `lib/lib.go` will be
+  a slice of nodes containing the root node, `lib` and finally `lib.go`
+  itself.
+
+- The children of each node can be lexicographically sorted by their name:
+  `LICENSE` will come before `lib`, that will come before
+  `main.go`.  Similarly, `lib.go` will always come before `lib_test.go`.
+
+- A tree depth-first traversal that visit nodes in the same directory in
+  lexicographic order will traverse the tree as the `tree` command line tool:
+
+  ```
+  LICENSE
+  lib
+  lib/lib.go
+  lib/lib_test.go
+  main.go
+  ```
+
+### Git trees as Merkle trees
+
+In a Git tree every node has a SHA-1 hash:
+
+- The hash of a file is calculated from its size and contents.
```

Comment:
Not fixed, I'm using the possessive pronoun here ("belongs to") not the contraction of "it is".